chia/cmds/param_types.py

```diff
@@ -0,0 +1,166 @@
+    """
+    A Click parameter type for transaction fees, which can be specified in XCH or mojos.
+    """
+
+    name: str = "uint64"  # output type
+    value_limit: Decimal = Decimal(0.5)
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> uint64:
+        if isinstance(value, uint64):  # required by click
+            return value
+        mojos = False  # TODO: Add unit logic
+        if mojos:
+            try:
+                return uint64(int(value))
+            except ValueError:
+                self.fail("Fee must be positive integer value in mojos", param, ctx)
+        try:
+            d_value = Decimal(value)
+            if 0 > d_value:
+                self.fail("Fee can not be negative", param, ctx)
+            if self.value_limit != Decimal(0) and d_value > self.value_limit:
+                self.fail(f"Fee must be in the range 0 to {self.value_limit}", param, ctx)
+            return uint64(int(d_value * units["chia"]))
+        except InvalidOperation:  # won't ever be value error because of the fee limit check
+            self.fail("Fee must be decimal dotted value in XCH (e.g. 0.00005)", param, ctx)
+
+
+@dataclass(frozen=True)
+class CliAmount:
+    """
+    A dataclass for TX / wallet amounts for both XCH and CAT, and of course mojos.
+    """
+
+    mojos: bool
+    amount: Union[uint64, Decimal]  # uint64 if mojos, Decimal if not
+
+    def convert_amount(self, mojo_per_unit: int) -> uint64:
+        if self.mojos:
+            if not isinstance(self.amount, uint64):
+                raise ValueError("Amount must be a uint64 if mojos flag is set.")
+            return self.amount
+        if not isinstance(self.amount, Decimal):
+            raise ValueError("Amount must be a Decimal if mojos flag is not set.")
+        return uint64(int(self.amount * mojo_per_unit))
+
+
+class AmountParamType(click.ParamType):
+    """
+    A Click parameter type for TX / wallet amounts for both XCH and CAT, and of course mojos.
+    """
+
+    name: str = "CliAmount"  # output type
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> CliAmount:
+        if isinstance(value, CliAmount):  # required by click
+            return value
+        if isinstance(value, uint64):
+            return CliAmount(mojos=True, amount=value)
+        mojos = False  # TODO: Add unit logic
+        if mojos:
+            try:
+                return CliAmount(mojos=True, amount=uint64(int(value)))
+            except ValueError:
+                self.fail("Amount must be positive integer value in mojos", param, ctx)
+        try:
+            if Decimal(value) <= 0:
+                self.fail("Amount must be greater than 0", param, ctx)
+            return CliAmount(mojos=False, amount=Decimal(value))
+        except InvalidOperation:
+            self.fail("Amount must be a decimal dotted value in XCH / CAT (e.g. 0.00005)", param, ctx)
```

Comment:
I would expect you to re-throw the exception here too

Reply:
view response above